### PR TITLE
fixed wrong casing color for alternative route lines and added tests

### LINKED ISSUE #3446: Alternative routes casing color regression
Found a regression from https://github.com/mapbox/mapbox-navigation-android/pull/3203 on the alternative routes casing color (gray vs black) - noting that we don't currently display alternative traffic congestions https://github.com/mapbox/mapbox-navigation-android/issues/3147 👀 

| Before https://github.com/mapbox/mapbox-navigation-android/pull/3203 | After https://github.com/mapbox/mapbox-navigation-android/pull/3203 |
|--------|--------------|
|   alternatives_routes_style_pre_3203     |        alternative_routes_style_post_3203      |

cc @cafesilencio @abhishek1508

### PR DESCRIPTION
**Pull request by cafesilencio**
## Description

fixes #3446 by restoring the alternative route casing color

- [ ] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

## Testing

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

**Comment by Guardiola31337** [comment]
Crash is easily reproducible, it happens at startup of any example in the test app.

**Comment by Guardiola31337** [comment]
> Crash is easily reproducible, it happens at startup of any example in the test app.

After latest push rebasing from `master` I can't reproduce anymore.

One thing that I noticed though is that interestingly the casing is working for alternatives but not for the primary route (refs. https://github.com/mapbox/mapbox-navigation-android/pull/3472) 👀 

casing_not_visible_in_primary_but_visible_in_alternatives

After discussing with @cafesilencio confirming that ☝️ is an issue with the gradient. As soon as you enable gradient the casing is not visible (for alternatives there's no gradient by default) so https://github.com/mapbox/mapbox-navigation-android/pull/3472 is still valid as a workaround until we get a fix from Maps.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProvider.kt [hunk 1]
```diff
@@ -44,27 +44,22 @@ import com.mapbox.navigation.ui.internal.route.RouteConstants.WAYPOINT_PROPERTY_
...
 import com.mapbox.navigation.ui.route.RouteStyleDescriptor
+import kotlin.reflect.KProperty1
 
 interface MapboxRouteLayerProvider : RouteLayerProvider {
     val routeStyleDescriptors: List<RouteStyleDescriptor>
 
-    fun getRouteLineColorExpressions(defaultColor: Int): List<Expression> {
+    fun getRouteLineColorExpressions(
+        defaultColor: Int,
+        routeColorProvider: KProperty1<RouteStyleDescriptor, Int>
```
Comment (Guardiola31337) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProvider.kt:54]: Do we really need `KProperty1`? `reflect` in general is scary. Also I'm wondering if that may introduce problems with Java compatibility as it's a Kotlin type 👀 at the ⚠️ https://kotlinlang.org/docs/reference/reflection.html#reflection

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapboxRouteLayerProviderFactory.kt [hunk 2]
```diff
@@ -1,12 +1,12 @@
-package com.mapbox.navigation.ui.internal.route
+package com.mapbox.navigation.ui.route
 
-import com.mapbox.navigation.ui.route.RouteStyleDescriptor
+import com.mapbox.navigation.ui.internal.route.MapboxRouteLayerProvider
```
Comment (Guardiola31337) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapboxRouteLayerProviderFactory.kt]: Can `MapboxRouteLayerProvider` be move out of the `internal` package and have the `internal` Kotlin modifier too?

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProviderFactoryTest.kt [hunk 3]
```diff
@@ -1,5 +1,6 @@
 package com.mapbox.navigation.ui.internal.route
 
+import com.mapbox.navigation.ui.route.MapboxRouteLayerProviderFactory
```
Comment (Guardiola31337) [inline libnavigation-ui/src/test/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProviderFactoryTest.kt:3]: Can `MapboxRouteLayerProviderFactoryTest` test be move out of the `internal` package too?

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProviderTest.kt [hunk 4]
```diff
@@ -1,5 +1,6 @@
 package com.mapbox.navigation.ui.internal.route
 
+import com.mapbox.navigation.ui.route.MapboxRouteLayerProvider
```
Comment (Guardiola31337) [inline libnavigation-ui/src/test/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProviderTest.kt:3]: Same as above re: move the test out of the `internal` package

Comment (cafesilencio) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProviderTest.kt]: done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProvider.kt (hunk 1 above)
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProvider.kt:54]: This saves from writing a lot of duplicate code in this case and the class is internal now so nobody external to the team will be accessing it.

Comment (Guardiola31337) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProvider.kt:54]: Ohh but we are as most of the UI codebase is in Java 😬

Comment (Guardiola31337) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/internal/route/MapboxRouteLayerProvider.kt:54]: Only wondering about the implications of depending on `reflect`.

**Pull request merged**